README.md
# FlowSorted.CordBloodCombined.450k
[![DOI](https://zenodo.org/badge/167406689.svg)](https://zenodo.org/badge/latestdoi/167406689)
This package includes a combination of four cell references for umbilical cord 
blood deconvolution using IlluminaHumanMethylation arrays 450K and EPIC 


License: GPL-3

The FlowSorted.CordBloodCombined.450k package contains data derived from 
Illumina HumanMethylation450K and Illumina HumanMethylationEPIC DNA methylation 
microarrays (Gervin K, Salas LA et al. 2018), consisting of 263 blood cell 
references and 26 umbilical cord blood samples, formatted as an RGChannelSet 
object for integration and normalization using most of the existing Bioconductor
packages. This is more than a combination of the original datasets, the package
has a rigorous curation that allows a better usage of the information.

This package contains cleaned data from four different umbilical cord blood 
references similar to the FlowSorted.CordBlood.450K package consisting of data 
from umbilical cord blood samples generated from healthy newborns. However, 
when using the cleaned dataset (eliminating potential cell cross contamination) 
and using the IDOL procedure compared to minfi estimates the cell type 
composition obtained through FlowSorted.CordBlood.450k package were less precise
and biased compared to actual cell counts. Hence, this package consists of 
appropriate data for deconvolution of umbilical cord blood samples used in for 
example EWAS relying in both 450K and EPIC technology.

Researchers may find this package useful as these samples represent different 
cellular populations ( T lymphocytes (CD4+ and CD8+), B cells (CD19+), monocytes
(CD14+), NK cells (CD56+) and Granulocytes of cell sorted umbilical cord blood. 
The estimates were contrasted versus FACS proportions in 22 samples, and 
validated using 197 umbilical cord blood samples. These data can be integrated 
with the minfi Bioconductor package to estimate cellular composition in users' 
umbilical cord blood Illumina 450K and EPIC samples using a modified version of 
the algorithm constrained projection/quadratic programming described in Houseman
et al. 2012. However, for more accurate estimations we suggests that the user 
prefers IDOL over minfi automatic estimations, using the function
estimateCellCounts2 from the package FlowSorted.Blood.EPIC which allows using 
customized sets of probes from IDOL.

References: 

KM Bakulski, et al. (2016) DNA methylation of cord blood 
cell types: Applications for mixed cell birth studies. Epigenetics 11:5. 
doi:10.1080/15592294.2016.1161875.

K Gervin, et al. (2016) Cell type specific DNA methylation in
cord blood: A 450K-reference data set and cell count-based validation of 
estimated cell type composition. Epigenetics 11:690–8. 
doi:10.1080/15592294.2016.1214782. 

OM de Goede, et al. (2015) Nucleated red blood cells impact DNA 
methylation and expression analyses of cord blood hematopoietic cells. 
Clin Epigenetics. 7:95. doi:10.1186/s13148-015-0129-6.

X Lin, et al. (2018) Cell type-specific DNA methylation in 
neonatal cord tissue and cord blood: A 850K-reference panel and comparison 
of cell-types. Epigenetics. 13:941–58. doi:10.1080/15592294.2018.1522929.

LA Salas et al. (2018). An optimized library for reference-based deconvolution 
of whole-blood biospecimens assayed using the  Illumina HumanMethylationEPIC 
BeadArray. Genome Biology Genome Biology 19, 64. doi: 10.1186/s13059-018-1448-7.

DC Koestler et al. (2016). Improving cell mixture deconvolution by identifying 
optimal DNA methylation libraries (IDOL). BMC bioinformatics. 17, 120.

EA Houseman et al. (2012) DNA methylation arrays as surrogate measures of cell 
mixture distribution. BMC Bioinformatics 13, 86.




inst/NEWS
Changes in version 0.99.0 (2019-01-21)
+ Submitted to Bioconductor
+ Added NEWS file
+ Initial development release

Changes in version 1.1.1 (2019-10-24)
+ Updated references
+ Fully compatible with estimateCellCounts2 in FlowSorted.Blood.EPIC



tests/testthat/test.R

#test_that("errors if bad parameters", {
#    library(FlowSorted.CordBloodCombined.450k)
#    library(ExperimentHub)
#    if (memory.limit()>8000){
#    hub <- ExperimentHub()
#    myfiles <- query(hub, "FlowSorted.CordBloodCombined.450k")
#    FlowSorted.CordBloodCombined.450k <- myfiles[[1]]
#    expect_false(!is(FlowSorted.CordBloodCombined.450k, "RGChannelSet"))
#    }
#})



inst/scripts/make-data.R
### =========================================================================
### FlowSorted.CordBloodCombined.450k 
### -------------------------------------------------------------------------
###
##Local processing
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(c("GEOquery", "minfi",
                       "FlowSorted.CordBlood.450k",
                       "FlowSorted.CordBloodNorway.450k"), version = "3.8")
getwd()
download.file(paste0("https://karnanilab.com/Tools/",
"FlowSorted.CordTissueAndBlood.EPIC/",
"FlowSorted.CordTissueAndBlood.EPIC_1.0.1.tar.gz"),
              destfile = paste0(getwd(), 
                            "/FlowSorted.CordTissueAndBlood.EPIC_1.0.1.tar.gz"))
install.packages(paste0(getwd(), 
                        "/FlowSorted.CordTissueAndBlood.EPIC_1.0.1.tar.gz"),
                 repos = NULL, type="source")
library(GEOquery)
library(minfi)
library(FlowSorted.CordBlood.450k)
library(FlowSorted.CordBloodNorway.450k)
library(FlowSorted.CordTissueAndBlood.EPIC)

#Prepare for combination
FlowSorted.CordBlood.450k<-FlowSorted.CordBlood.450k
FlowSorted.CordBlood.450k@colData[] <- lapply(FlowSorted.CordBlood.450k@colData,
                                              function(x) if (is.factor(x)) 
                                                  as.character(x) else {x})
FlowSorted.CordBlood.450k@colData[] <- lapply(FlowSorted.CordBlood.450k@colData,
                                              function(x) if (is.logical(x)) 
                                                  as.character(x) else {x})
FlowSorted.CordBlood.450k$Study<-"Bakulski"

FlowSorted.CordBloodNorway.450k<-FlowSorted.CordBloodNorway.450k
FlowSorted.CordBloodNorway.450k@colData[] <- 
    lapply(FlowSorted.CordBloodNorway.450k@colData, function(x) 
        if (is.factor(x)) as.character(x) else {x})
FlowSorted.CordBloodNorway.450k@colData[] <- 
    lapply(FlowSorted.CordBloodNorway.450k@colData, function(x) 
        if (is.logical(x)) as.character(x) else {x})
FlowSorted.CordBloodNorway.450k$Study<-"Gervin"

FlowSorted.CordTissueAndBlood.EPIC<-updateObject(
    FlowSorted.CordTissueAndBlood.EPIC)
FlowSorted.CordTissueAndBlood.EPIC@colData[] <- lapply(
    FlowSorted.CordTissueAndBlood.EPIC@colData, function(x) if (is.factor(x)) 
        as.character(x) else {x})
FlowSorted.CordTissueAndBlood.EPIC@colData[] <- lapply(
    FlowSorted.CordTissueAndBlood.EPIC@colData, function(x) if (is.logical(x)) 
        as.character(x) else {x})
FlowSorted.CordTissueAndBlood.EPIC$Study<-"Lin"

getGEOSuppFiles("GSE68456")
untar("GSE68456/GSE68456_RAW.tar", exdir = "GSE68456/idat")
idatFiles <- list.files("GSE68456/idat", pattern = "idat.gz$", full = TRUE)
sapply(idatFiles, gunzip, overwrite = TRUE)
library(FlowSorted.CordBloodCanada.450k)
FlowSorted.CordBloodCanada.450k <- read.metharray.exp("GSE68456/idat")
geoMat <- getGEO("GSE68456")
pD <- pData(geoMat[[1]])
sampleNames(FlowSorted.CordBloodCanada.450k) <- substr(
    sampleNames(FlowSorted.CordBloodCanada.450k),1, 10)
rownames(pD) <- pD$geo_accession
pD <- pD[sampleNames(FlowSorted.CordBloodCanada.450k),]
pData(FlowSorted.CordBloodCanada.450k) <- DataFrame(pD, 
                                                    row.names = rownames(pD))
FlowSorted.CordBloodCanada.450k

FlowSorted.CordBloodCanada.450k$Study<-"deGoede"
FlowSorted.CordBloodCanada.450k$Sex<-FlowSorted.CordBloodCanada.450k$Sex.ch1
FlowSorted.CordBloodCanada.450k$CellType<-NA
FlowSorted.CordBloodCanada.450k$CellType<-as.character(
    FlowSorted.CordBloodCanada.450k$CellType)
levels(FlowSorted.CordBloodCanada.450k$characteristics_ch1.1)
FlowSorted.CordBloodCanada.450k$CellTypeLong<-ifelse(
    FlowSorted.CordBloodCanada.450k$characteristics_ch1.1=="cell type: B cell",
    "Bcell",
    ifelse(FlowSorted.CordBloodCanada.450k$characteristics_ch1.1==
               "cell type: CD4 T cell", "CD4T",
           ifelse(FlowSorted.CordBloodCanada.450k$characteristics_ch1.1==
                      "cell type: CD8 T cell", "CD8T",
                ifelse(FlowSorted.CordBloodCanada.450k$characteristics_ch1.1==
                           "cell type: Granulocyte", "Gran",
                ifelse(FlowSorted.CordBloodCanada.450k$characteristics_ch1.1==
                           "cell type: Monocyte", "Mono",
                ifelse(FlowSorted.CordBloodCanada.450k$characteristics_ch1.1==
                           "cell type: NK cell", "NK", 
                ifelse(FlowSorted.CordBloodCanada.450k$characteristics_ch1.1==
                           "cell type: T cell", "PanT", "nRBC")))))))   
FlowSorted.CordBloodCanada.450k$CellType<-as.character(
    FlowSorted.CordBloodCanada.450k$CellTypeLong)
FlowSorted.CordBloodCanada.450k$CellType2<-
    FlowSorted.CordBloodCanada.450k$CellType
FlowSorted.CordBloodCanada.450k@colData[] <- lapply(
    FlowSorted.CordBloodCanada.450k@colData, function(x) if (is.factor(x)) 
        as.character(x) else {x})
FlowSorted.CordBloodCanada.450k@colData[] <- lapply(
    FlowSorted.CordBloodCanada.450k@colData, function(x) if (is.logical(x)) 
        as.character(x) else {x})


tmp<-combineArrays(FlowSorted.CordBlood.450k,FlowSorted.CordBloodNorway.450k)
tmp<-combineArrays(tmp,FlowSorted.CordBloodCanada.450k)
tmp<-combineArrays(tmp,FlowSorted.CordTissueAndBlood.EPIC)

#read the phenotype data available in the inst/extdata
pheno<-read.csv(file="inst/extdata/Pheno_combined.csv", header=TRUE, 
                row.names = 1)

FlowSorted.CordBloodCombined.450k<-tmp[, rownames(pheno)]
identical(colnames(FlowSorted.CordBloodCombined.450k), rownames(pheno))
pData(FlowSorted.CordBloodCombined.450k)<-DataFrame(pheno, 
                                                    row.names = rownames(pheno))
FlowSorted.CordBloodCombined.450k@colData[] <- lapply(
    FlowSorted.CordBloodCombined.450k@colData, function(x) if (is.factor(x)) 
        as.character(x) else {x})
FlowSorted.CordBloodCombined.450k@colData[] <- lapply(
    FlowSorted.CordBloodCombined.450k@colData, function(x) if (is.logical(x)) 
        as.character(x) else {x})


save(FlowSorted.CordBloodCombined.450k, 
     file="data/FlowSorted.CordBloodCombined.450k.rda", version = 3,
     compress = TRUE)




inst/scripts/make-metadata.R
### =========================================================================
### FlowSorted.CordBloodCombined.450k metadata 
### -------------------------------------------------------------------------
###

meta <- data.frame(
    Title = c("FlowSorted.CordBloodCombined.450k"),
    Description = c(paste0("The FlowSorted.CordBloodCombined.450k package ",
                           "contains data derived from Illumina ", 
                           "HumanMethylation450K and ", 
                           "Illumina HumanMethylationEPIC ",
                           "DNA methylation microarrays (Gervin K, Salas LA ",
                           "et al. under review), consisting of 263 blood ",
                           "cell references and 26 umbilical cord blood ",
                           "samples, formatted as an RGChannelSet object for ",
                           "integration and normalization using most of the ",
                           "existing Bioconductor packages.")),
    BiocVersion = c("3.9"),
    Genome = rep("hg19", 1), 
    SourceType = rep("tar.gz", 1), 
    SourceUrl = paste0("https://doi.org/doi:10.18129/",
                         "B9.bioc.FlowSorted.CordBlood.450k , ",
                  "https://doi.org/doi:10.18129/",
                         "B9.bioc.FlowSorted.CordBloodNorway.450k , ",
        "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE68456 , ",
                  "https://karnanilab.com/Tools/",
                         "FlowSorted.CordTissueAndBlood.EPIC/",
                         "FlowSorted.CordTissueAndBlood.EPIC_1.0.1.tar.gz"),
    SourceVersion = "Jan 18 2019",
    Species = "Homo sapiens",
    TaxonomyId = 9606,
    Coordinate_1_based = TRUE,
    DataProvider = paste0("Bioconductor, ", "Bioconductor, ","GEO, ",
                          "karnanilab"),
    Maintainer = "Lucas A Salas <lucas.a.salas.diaz@dartmouth.edu>",
    RDataClass = c("RGChannelSet") ,
    DispatchClass = c(rep("Rda",1)),
RDataPath = c(paste0("FlowSorted.CordBloodCombined.450k/",
"FlowSorted.CordBloodCombined.450k.rda")),
Tags = "",
Notes = paste0("Bakulski et al 2016 data, ", "Gervin et al 2016 data, ",
          "de Goede et al 2015 data, ", "Lin et al 2018 data")
)

write.csv(meta, file="inst/extdata/metadata.csv", row.names=FALSE)


R/IDOLOptimizedCpGsCordBlood.R
#'     IDOL Optimized CpGs for umbilical cord blood DNA methylation 
#'     deconvolution 
#' 
#' @description 
#'     This object is a vector of length 517 consisting of the names of the IDOL
#'     optimized CpGs.  These CpGs are used as the backbone for deconvolution
#'     and were selected because their methylation signature differs across the 
#'     six normal leukocyte subtypes and the nucleated red blood cells.
#' 
#' 
#' @format An object of class "character" of length 517.
#' 
#'         The format is:
#'         chr [1:517] "cg12603453" "cg24765783" "cg06975018" "cg19708055" ...
#' 
#' @references K Gervin, LA Salas et al. (2019) \emph{Systematic evaluation and 
#' validation of references and library selection methods for deconvolution of 
#' cord blood DNA methylation data}. Clin Epigenetics 11,125. doi:
#' 10.1186/s13148-019-0717-y
#' @references LA Salas et al. (2018). \emph{An optimized library for 
#' reference-based deconvolution of whole-blood biospecimens assayed using the 
#' Illumina HumanMethylationEPIC BeadArray}. Genome Biology 19, 64. doi:
#' 10.1186/s13059-018-1448-7.
#' @references DC Koestler et al. (2016). \emph{Improving cell mixture 
#' deconvolution by identifying optimal DNA methylation libraries (IDOL)}. 
#' BMC bioinformatics. 17, 120. doi: 10.1186/s12859-016-0943-7.
#' 
#' @examples
#' # Do not run
#' # library (FlowSorted.Blood.EPIC)
#' # data("IDOLOptimizedCpGsCordBlood")
#' # Step 1: Load the reference library to extract the umbilical cord samples
#' # library(ExperimentHub)
#' # hub <- ExperimentHub()
#' # myfiles <- query(hub, "FlowSorted.CordBloodCombined.450k")
#' # FlowSorted.CordBloodCombined.450k <- myfiles[[1]]
#' # FlowSorted.CordBloodCombined.450k
#' 
#' # Step 2 separate the reference from the testing dataset if you want to run 
#' # examples for estimations for this function example
#' 
#' # RGsetTargets <- FlowSorted.CordBloodCombined.450k[,
#' # FlowSorted.CordBloodCombined.450k$CellType == "WBC"]
#' # sampleNames(RGsetTargets) <- paste(RGsetTargets$CellType,
#' #                               seq_len(dim(RGsetTargets)[2]), sep = "_")
#' # RGsetTargets
#' 
#' # Step 3: use your favorite package for deconvolution.
#' # Deconvolute a target data set consisting of 450K DNA methylation 
#' # data profiled in blood, using your prefered method.
#' # You can use our IDOL optimized DMR library for the Cord Blood,  This object
#' # contains a vector of length 517 consisting of the IDs of the IDOL optimized
#' # CpG probes.  These CpGs are used as the backbone for deconvolution and were
#' # selected because their methylation signature differs across the six normal 
#' # leukocyte subtypes plus the nucleated red blood cells.
#' 
#' # data (IDOLOptimizedCpGsCordBlood)
#' # head (IDOLOptimizedCpGsCordBlood)
#' # We recommend using Noob processMethod = "preprocessNoob" in minfi for the 
#' # target and reference datasets. 
#' # Cell types included are "CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran", 
#' # "nRBC"
#' # To use the IDOL optimized list of CpGs (IDOLOptimizedCpGsCordBlood) use 
#' # estimateCellCounts2 from FlowSorted.Blood.EPIC. 
#' # Do not run with limited RAM the normalization step requires a big amount 
#' # of memory resources. Use the parameters as specified below for 
#' # reproducibility.
#' # 
#' # if (memory.limit()>8000){
#' #     countsUCB<-estimateCellCounts2(RGsetTargets, 
#' #                                     compositeCellType = 
#' #                                                "CordBloodCombined", 
#' #                                     processMethod = "preprocessNoob",
#' #                                     probeSelect = "IDOL", 
#' #                                     cellTypes = c("CD8T", "CD4T", "NK",  
#' #                                     "Bcell", "Mono", "Gran", "nRBC"), 
#' #                                     referencePlatform = 
#' #                                         "IlluminaHumanMethylation450k",
#' #                                     referenceset = 
#' #                                      "FlowSorted.CordBloodCombined.450k",
#' #                                     IDOLOptimizedCpGs =
#' #                                       IDOLOptimizedCpGsCordBlood, 
#' #                                     returnAll = FALSE)
#' #     
#' #     head(countsUCB$counts)
#' # }
"IDOLOptimizedCpGsCordBlood"


R/FlowSorted.CordBloodCombined.450k.R
#' FlowSorted.CordBloodCombined.450k
#' @description
#' The FlowSorted.CordBloodCombined.450k package contains data derived from 
#' Illumina HumanMethylation450K and Illumina HumanMethylationEPIC DNA 
#' methylation microarrays (Gervin K, Salas LA et al. under review), consisting
#' of 263 blood cell references and 26 umbilical cord blood samples, formatted 
#' as an RGChannelSet object for integration and normalization using most of  
#' the existing Bioconductor packages.
#' 
#' This package contains cleaned data from four different umbilical cord blood 
#' references similar to the FlowSorted.CordBlood.450K package consisting of 
#' data from umbilical cord blood samples generated from healthy newborns. 
#' However, when using the cleaned dataset (eliminating potential cell cross 
#' contamination) and using the IDOL procedure compared to minfi estimates the 
#' cell type composition obtained through FlowSorted.CordBlood.450k package were
#' less precise and biased compared to actual cell counts. Hence, this package 
#' consists of appropriate data for deconvolution of umbilical cord blood 
#' samples used in for example EWAS relying in both 450K and EPIC technology.
#' 
#' Researchers may find this package useful as these samples represent  
#' different cellular populations ( T lymphocytes (CD4+ and CD8+), B cells  
#' (CD19+), monocytes (CD14+), NK cells (CD56+), Granulocytes, and nucleated 
#' red blood cells of cell sorted umbilicalcord blood. The estimates were 
#' contrasted versus FACS proportions in 22 umbilical samples, and validated 
#' using 197 umbilical cord blood samples.  
#' 
#' These data can be integrated with the minfi Bioconductor package to   
#' estimate cellular composition in users' umbilical cord blood Illumina 450K  
#' and EPIC samples using a modified version of the algorithm constrained  
#' projection/quadratic programming described in Houseman et al. 2012. However, 
#' for more accurate estimations we suggests that the user prefers IDOL over  
#' minfi automatic estimations, using the function estimateCellCounts2 from   
#' the package FlowSorted.Blood.EPIC which allows using customized sets of   
#' probes from IDOL (see IDOLOptimizedCpGsCordBlood for an example).
#'
#' @import minfi
#' @import SummarizedExperiment
#' @import IlluminaHumanMethylation450kanno.ilmn12.hg19
#' @import IlluminaHumanMethylationEPICanno.ilm10b4.hg19
#' @import ExperimentHub
#' @import FlowSorted.Blood.EPIC
#' @importFrom utils data
#' @importFrom utils read.csv
#' @importFrom utils memory.limit

#' 
#' @format A class: RGChannelSet, dimensions: 575130 289 
#' 
#' @seealso
#' References \enumerate{
#' \item  K Gervin, LA Salas et al. (2019) \emph{Systematic evaluation and 
#' validation of references and library selection methods for deconvolution of 
#' cord blood DNA methylation data}. Clin Epigenetics 11,125. doi:
#' 10.1186/s13148-019-0717-y
#' \item  KM Bakulski, et al. (2016) \emph{DNA methylation of cord blood 
#' cell types: Applications for mixed cell birth studies}. Epigenetics 11:5. 
#' doi:10.1080/15592294.2016.1161875.
#' \item  K Gervin, et al. (2016) \emph{Cell type specific DNA methylation in
#' cord blood: A 450K-reference data set and cell count-based validation of 
#' estimated cell type composition}. Epigenetics 11:690-8. 
#' doi:10.1080/15592294.2016.1214782. 
#' \item  OM de Goede, et al. (2015) \emph{Nucleated red blood cells impact DNA 
#' methylation and expression analyses of cord blood hematopoietic cells}. 
#' Clin Epigenetics. 7:95. doi:10.1186/s13148-015-0129-6.
#' \item  X Lin, et al. (2018) \emph{Cell type-specific DNA methylation in 
#' neonatal cord tissue and cord blood: A 850K-reference panel and comparison 
#' of cell-types}. Epigenetics. 13:941-58. doi:10.1080/15592294.2018.1522929.
#' \item LA Salas et al. (2018). \emph{An optimized library for 
#' reference-based deconvolution of whole-blood biospecimens assayed using the 
#' Illumina HumanMethylationEPIC BeadArray}. Genome Biology 19, 64. doi:
#' 10.1186/s13059-018-1448-7.
#' \item DC Koestler et al. (2016). \emph{Improving cell mixture deconvolution 
#' by identifying optimal DNA methylation libraries (IDOL)}. BMC bioinformatics.
#' 17, 120. doi: 10.1186/s12859-016-0943-7.
#' \item EA Houseman et al. (2012) \emph{DNA methylation arrays as surrogate
#' measures of cell mixture distribution}. BMC Bioinformatics 13, 86.
#' doi:10.1186/1471-2105-13-86.
#' \item \pkg{minfi} package for tools for estimating cell type
#' composition in blood using these data
#' }
#' 
#' @examples
#' library(ExperimentHub)
#' hub <- ExperimentHub()
#' myfiles <- query(hub, "FlowSorted.CordBloodCombined.450k")
#' myfiles$ah_id=="EH2256"#ExperimenHub ID
#' # Do not run unless you have enough memory resources 
#' # FlowSorted.CordBloodCombined.450k <- myfiles[[1]]
#' # FlowSorted.CordBloodCombined.450k
#' 
#' @return RGChannelSet 289 samples
#' @usage 
#' FlowSorted.CordBloodCombined.450k 
"FlowSorted.CordBloodCombined.450k"
